### ISSUE TITLE
javax.ws.rs.ProcessingException: java.net.BindException: Address already in use: connect

### ISSUE THREAD

**Issue by YuriiMalskyi** [EMResearch/EvoMaster#213]
00:28:00.886 [main] WARN  o.e.c.p.r.s.AbstractRestFitness - Running out of ephemeral ports. Waiting 30 seconds before re-trying connection
* [ERROR] EvoMaster process terminated abruptly. This is likely a bug in EvoMaster. Please copy&paste the following stacktrace, and create a new issue on https://github.com/EMResearch/EvoMaster/issues
javax.ws.rs.ProcessingException: java.net.BindException: Address already in use: connect
        at org.glassfish.jersey.client.internal.HttpUrlConnector.apply(HttpUrlConnector.java:260)
        at org.glassfish.jersey.client.ClientRuntime.invoke(ClientRuntime.java:254)
        at org.glassfish.jersey.client.JerseyInvocation.lambda$invoke$0(JerseyInvocation.java:736)
        at org.glassfish.jersey.internal.Errors.process(Errors.java:292)
        at org.glassfish.jersey.internal.Errors.process(Errors.java:274)
        at org.glassfish.jersey.internal.Errors.process(Errors.java:205)
        at org.glassfish.jersey.process.internal.RequestScope.runInScope(RequestScope.java:390)
        at org.glassfish.jersey.client.JerseyInvocation.invoke(JerseyInvocation.java:735)
        at org.evomaster.core.problem.rest.service.AbstractRestFitness.handleRestCall(AbstractRestFitness.kt:343)
        at org.evomaster.core.problem.rest.service.BlackBoxRestFitness.doCalculateCoverage(BlackBoxRestFitness.kt:53)
        at org.evomaster.core.problem.rest.service.BlackBoxRestFitness.doCalculateCoverage(BlackBoxRestFitness.kt:16)
        at org.evomaster.core.search.service.FitnessFunction.calculateCoverage(FitnessFunction.kt:46)
        at org.evomaster.core.search.algorithms.RandomAlgorithm.search(RandomAlgorithm.kt:25)
        at org.evomaster.core.Main$Companion.run(Main.kt:280)
        at org.evomaster.core.Main$Companion.initAndRun(Main.kt:148)
        at org.evomaster.core.Main$Companion.main(Main.kt:72)
        at org.evomaster.core.Main.main(Main.kt)
Caused by: java.net.BindException: Address already in use: connect
        at java.net.DualStackPlainSocketImpl.connect0(Native Method)
        at java.net.DualStackPlainSocketImpl.socketConnect(Unknown Source)
        at java.net.AbstractPlainSocketImpl.doConnect(Unknown Source)
        at java.net.AbstractPlainSocketImpl.connectToAddress(Unknown Source)
        at java.net.AbstractPlainSocketImpl.connect(Unknown Source)
        at java.net.PlainSocketImpl.connect(Unknown Source)
        at java.net.SocksSocketImpl.connect(Unknown Source)
        at java.net.Socket.connect(Unknown Source)
        at sun.net.NetworkClient.doConnect(Unknown Source)
        at sun.net.www.http.HttpClient.openServer(Unknown Source)
        at sun.net.www.http.HttpClient.openServer(Unknown Source)
        at sun.net.www.http.HttpClient.<init>(Unknown Source)
        at sun.net.www.http.HttpClient.New(Unknown Source)
        at sun.net.www.http.HttpClient.New(Unknown Source)
        at sun.net.www.protocol.http.HttpURLConnection.getNewHttpClient(Unknown Source)
        at sun.net.www.protocol.http.HttpURLConnection.plainConnect0(Unknown Source)
        at sun.net.www.protocol.http.HttpURLConnection.plainConnect(Unknown Source)
        at sun.net.www.protocol.http.HttpURLConnection.connect(Unknown Source)
        at sun.net.www.protocol.http.HttpURLConnection.getOutputStream0(Unknown Source)
        at sun.net.www.protocol.http.HttpURLConnection.getOutputStream(Unknown Source)
        at org.glassfish.jersey.client.internal.HttpUrlConnector.lambda$_apply$0(HttpUrlConnector.java:358)
        at org.glassfish.jersey.message.internal.CommittingOutputStream.commitStream(CommittingOutputStream.java:171)
        at org.glassfish.jersey.message.internal.CommittingOutputStream.commitStream(CommittingOutputStream.java:165)
        at org.glassfish.jersey.message.internal.CommittingOutputStream.write(CommittingOutputStream.java:199)
        at org.glassfish.jersey.message.internal.WriterInterceptorExecutor$UnCloseableOutputStream.write(WriterInterceptorExecutor.java:276)
        at sun.nio.cs.StreamEncoder.writeBytes(Unknown Source)
        at sun.nio.cs.StreamEncoder.implFlushBuffer(Unknown Source)
        at sun.nio.cs.StreamEncoder.implFlush(Unknown Source)
        at sun.nio.cs.StreamEncoder.flush(Unknown Source)
        at java.io.OutputStreamWriter.flush(Unknown Source)
        at org.glassfish.jersey.message.internal.ReaderWriter.writeToAsString(ReaderWriter.java:170)
        at org.glassfish.jersey.message.internal.AbstractMessageReaderWriterProvider.writeToAsString(AbstractMessageReaderWriterProvider.java:107)
        at org.glassfish.jersey.message.internal.StringMessageProvider.writeTo(StringMessageProvider.java:76)
        at org.glassfish.jersey.message.internal.StringMessageProvider.writeTo(StringMessageProvider.java:36)
        at org.glassfish.jersey.message.internal.WriterInterceptorExecutor$TerminalWriterInterceptor.invokeWriteTo(WriterInterceptorExecutor.java:242)
        at org.glassfish.jersey.message.internal.WriterInterceptorExecutor$TerminalWriterInterceptor.aroundWriteTo(WriterInterceptorExecutor.java:227)
        at org.glassfish.jersey.message.internal.WriterInterceptorExecutor.proceed(WriterInterceptorExecutor.java:139)
        at org.glassfish.jersey.message.internal.MessageBodyFactory.writeTo(MessageBodyFactory.java:1115)
        at org.glassfish.jersey.client.ClientRequest.doWriteEntity(ClientRequest.java:503)
        at org.glassfish.jersey.client.ClientRequest.writeEntity(ClientRequest.java:485)
        at org.glassfish.jersey.client.internal.HttpUrlConnector._apply(HttpUrlConnector.java:360)
        at org.glassfish.jersey.client.internal.HttpUrlConnector.apply(HttpUrlConnector.java:258)
        ... 16 common frames omitted

**Comment by arcuri82** [comment]
Hi,
 it looks like you are running out of ephemeral ports on your OS (ie, it is not possible to establish new TCP connections as the OS is not freeing them fast enough).
Can you provide a bit more details on how you are running EvoMaster (EM)? eg:
- you are using BlackBox, isn't it?
- do you have a single EM instance? or running several in parallel? 
- which OS? (if it is Windows, might want to look at the documentation for [Windows and networking](https://github.com/EMResearch/EvoMaster/blob/master/docs/troubleshooting/windows.md))

**Comment by YuriiMalskyi** [comment]
Hello, yes i'm using a BlackBox testing, and I was running several of them in parallel (3). Looks like that was the issue, as during the wait time the ports could already be filled with another instance of this "service/tool". 
If I run only one - everything is perfect.

P.S. Thanks for a quick answer and sorry for disturbing

**Comment by arcuri82** [comment]
np ;)
